keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+type IPRangeRef struct {
+	// APIGroup is the group for the resource being referenced
+	APIGroup string
+	// Kind is the type of resource being referenced
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
+	// Address is the text representation of the IP Address.
+	Address string `json:"address"`
+	// IPRangeRef references the IPRange associated to this IP Address.
+	IPRangeRef IPRangeRef `json:"ipRangeRef,omitempty"`
+}
+
+// IPAddressStatus defines the observed state of IPAddress.
+type IPAddressStatus struct {
+	State IPAddressState `json:"state,omitempty"`
+}
+
+// IPAddressState defines the state of the IP address
+type IPAddressState string
+
+// These are the valid statuses of IPAddresses.
+const (
+	// IPPending means the IP has been allocated by the system but the object associated
+	// (typically Services ClusterIPs) has not been persisted yet.
+	IPPending IPAddressState = "Pending"
+	// IPAllocated means the IP has been persisted with the object associated.
+  IPAllocated IPAddressState = "Allocated"
+  // IPReserved means the IP has been reserved and can not be allocated to any Service.
+	IPReserved IPAddressState = "Reserved"
+	// IPFree means that IP has not been allocated neither persisted.
+	IPFree IPAddressState = "Free"
+)
+```
+
+The name of the IPAddress has to be unique, so we can use it to provide consistency and avoid duplicate allocations.
+However, Kubernetes objects doesn't allow names with `:`, used by IPv6 Addresses. In addition, [IPv6 addresses
+can have different text representations](https://tools.ietf.org/html/rfc4291).
+
+To guarantee the name uniqueness, we can use the decimal representation of the IP address, since IP can be a 32 bits or a 128 bits number. To facilitate this operations we introduce two new utils functions:
+
+```go
+// IPToDecimal converts an IP to a string with its decimal representation
+func IPToDecimal(ip net.IP) string {
+	if ip == nil {
+		return ""
+	}
+	i := new(big.Int)
+	// compare the address with its decimal representation
+	if ip.To4() != nil {
+		i.SetBytes(ip.To4())
+	} else if ip.To16() != nil {
+		i.SetBytes(ip.To16())
+	}
+	return i.String()
+}
+
+// DecimalToIP converts a string with the decimal representation of an IP address and returns the IP
+func DecimalToIP(ipDecimal string) net.IP {
+	if len(ipDecimal) == 0 {
+		return nil
+	}
+	// convert decimal representation to BigInt
+	i, ok := new(big.Int).SetString(ipDecimal, 10)
+	// if we can not convert the string to a number
+	// or the number has more than 16 bytes it is not an IP
+	if !ok || len(i.Bytes()) > 16 {
+		return nil
+	}
+	bufLen := net.IPv4len
+	if len(i.Bytes()) > net.IPv4len {
+		bufLen = net.IPv6len
+	}
+	// convert BigInt to IP address, truncate to IP Len bytes
+	r := append(make([]byte, bufLen), i.Bytes()...)
+	return net.IP(r[len(r)-bufLen:])
+}
+```
```

Comment:
It seems I didn't understand well the `name` requirements and we can use normal ips as names, so this is no needed
https://github.com/kubernetes/enhancements/pull/1881#issuecomment-766557739

Reply:
It *must* be canonicalized somehow, though. It's much trickier than it should be: https://blog.dave.tf/post/ip-addr-parsing/

Reply:
yep, the problem is IPv6, if we don't use the 32bits or 128bits integer that I used here, I guess that we should use something like https://tools.ietf.org/html/rfc5280#section-4.2.1.6 representing the not abbreviated from to avoid ambiguity

2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b

Reply:
yeah, both v4 and v6 have canonical formats.  We can just use those.

Reply:
To be clear, https://tools.ietf.org/html/rfc5280#section-4.2.1.6 is a _binary_ format. For a canonical IPv6 string format see  https://tools.ietf.org/html/rfc5952#section-4.0 (IPv6); there isn't an RFC covering that for IPv4, but the dotted-decimal format is widely treated as canonical.

`2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b` is noncanonical as it does not use `::` as defined in RFC5280.
(this also reminds me that I have let https://github.com/dspinhirne/netaddr-go/pull/2 go stale).

Reply:
I see, so I got it wrong, thanks @sftim for clarifying

Reply:
2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b is not _wrong_ as a representation. It's valid, and easier to teach than the canonical form that is a SHOULD in https://tools.ietf.org/html/rfc5952#section-4.0

Reply:
The thing is that name has to be unique, and I foresee people complaining why their valid ips are not accepted, so I think that the long format doesn't leave room to interpretation ...

Reply:
I think we could implement expansion (or contraction) to the canonical format as a mutating admission webhook - eg 2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b in the API, but allow `2001:db8:3c4d:15::1a2f:1a2b` in manifests. Or vice versa.

That can be left for future improvement right? It doesn't block progress on the allocators API AIUI.

Reply:
more on this topic, seems everybody is having the same problem https://github.com/golang/go/issues/30264

the apiserver has a hook after validaton called Canonicalize, I think that this is what TimHockin and Daniel are referring to
```
// Canonicalize normalizes the object after validation.
--
123 | func (svcStrategy) Canonicalize(obj runtime.Object) {
```

Reply:
My (unstated!) point with that link was that you can encode an IPv6-formatted address that is really an IPv4 address. Whatever canonicalization strategy we use has to account for that.

Also note that cannonicalizing the name is a little different than putting other parts of the object into a standard form. With the name, you just have to reject the request if the user uses a non-canonical name, you can't fix it up for them.

Reply:
the name can be the canonical name, that simplifies everything